proto/src/error.rs

```diff
@@ -0,0 +1,131 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum LdapProtoError {
```

Comment:
Listing all potential errors seems a bit brittle, maybe?

What about some categories, like "invalid BER" with a string for the details? Or are you specifically trying to avoid String/allocations?
At the very least I would make this enum non-exhaustive, so adding a new error is not a breaking change.

Reply:
Also, do we use this anywhere?

Reply:
Yeah, I'm avoiding string allocs here. The idea is each error encodes "what and where" it occured. It's kind of a "learnt lesson" from kanidm where we used generic errors, people would say "we had X error" and then there are a few hundreds places it could arise from. So we are moving to more precise errors in Kanidm, and an example would be in kani to wrap this as:

`LdapE001 ( LdapProtoError )`

Which then tells us precisely where the error was, and in this case, also extends to tell us more about what Ldap error occured too. This also then has implications for internationalisation because then we can have translations later since we avoided human strings. It also saves space in serialisation/encoding since we could choose to repr this as u16 or similar instead of strings in Json. So generally this is "forward looking" to a large error rework with Kanidm.

Reply:
You don't want to at least make the enum non exhaustive?